### PR TITLE
Allow nilable Spoom::Location lines and columns attributes

### PR DESCRIPTION
**Pull request by alexcrocha**
## Summary
This PR updates the `Spoom::Location` class to allow the `start_line`, `end_line`, `start_column`, and `end_column` attributes to be `nilable`. This change provides greater flexibility in representing different types of locations within a file.

## Details
- **Nilable Attributes**: The `start_line`, `end_line`, `start_column`, and `end_column` attributes of `Spoom::Location` can now be nil.
- **Flexible `Spoom::Location` Representation**:
   - Can now represent an entire file, e.g., `foo.rb`.
   - It can also represent a range of lines within a file, e.g., `foo.rb:1-3`.
   - Maintains previous implementation, with both lines and columns, e.g., `foo.rb:1:2-3:4`.
   
## What should Reviewers pay attention to
- **Exception Messages**: Added more detailed messages to exceptions raised within the `Spoom::Location` class to improve debugging and error handling. Unsure if this is a good change.
- **`Float::Infinity`**: Introduced `Float::Infinity` to represent inclusivity when lines or columns are nil, ensuring that ranges can be handled correctly even when the start or end is not explicitly known.

### CODE REVIEW

**Comment by andyw8** [comment]
Are there perhaps two distinct concepts here that should be seperated out, e.g. `LocationPosition` and a `LocationRange` ?

File: lib/spoom/location.rb [hunk 1]
```diff
@@ -15,42 +15,77 @@ class << self
       sig { params(location_string: String).returns(Location) }
       def from_string(location_string)
         file, rest = location_string.split(":", 2)
-        raise LocationError, "Invalid location string: #{location_string}" unless file && rest
+        raise LocationError, "Invalid location. File name is missing in location string: #{location_string}" unless file
```
Comment (Morriar) [inline lib/spoom/location.rb:18]: Let's reword a bit the error messages:

```suggestion
        raise LocationError, "Invalid location string `#{location_string}`: missing file name" unless file
```

File: lib/spoom/location.rb [hunk 2]
```diff
@@ -15,42 +15,77 @@ class << self
...
+
+        return new(file) if rest.nil?
 
         start_line, rest = rest.split(":", 2)
-        raise LocationError, "Invalid location string: #{location_string}" unless start_line && rest
+        if rest.nil?
+          start_line, end_line = T.must(start_line).split("-", 2)
+          return new(file, start_line: start_line.to_i, end_line: end_line.to_i) if end_line
+
+          raise LocationError, "Invalid location. End line is missing in range format: #{location_string}"
```
Comment (Morriar) [inline lib/spoom/location.rb:27]:
```suggestion
          raise LocationError, "Invalid location string `#{location_string}`: missing end line"
```

Comment (Morriar) [inline lib/spoom/location.rb:27]: I'd prefer an early raise and an unconditional return:

```suggestion
          start_line, end_line = T.must(start_line).split("-", 2)
          raise LocationError, "Invalid location. End line is missing in range format: #{location_string}" unless end_line
          
          return new(file, start_line: start_line.to_i, end_line: end_line.to_i) if end_line
```

So the style and logic is coherent with the other branches.

File: lib/spoom/location.rb [hunk 3]
```diff
@@ -15,42 +15,77 @@ class << self
...
+        end
 
         start_column, rest = rest.split("-", 2)
-        raise LocationError, "Invalid location string: #{location_string}" unless start_column && rest
+        if rest.nil?
+          raise LocationError,
+            "Invalid location. End line and column are missing in location string: #{location_string}"
```
Comment (Morriar) [inline lib/spoom/location.rb:33]:
```suggestion
            "Invalid location string `#{location_string}`: missing end line and column"
```

File: lib/spoom/location.rb [hunk 4]
```diff
@@ -15,42 +15,77 @@ class << self
...
+        end
 
         end_line, end_column = rest.split(":", 2)
-        raise LocationError, "Invalid location string: #{location_string}" unless end_line && end_column
-
-        new(file, start_line.to_i, start_column.to_i, end_line.to_i, end_column.to_i)
+        unless end_line && end_column
+          raise LocationError, "Invalid location. End column is missing in location string: #{location_string}"
+        end
```
Comment (Morriar) [inline lib/spoom/location.rb:39]:
```suggestion
        raise LocationError, "Invalid location string `#{location_string}`: missing end column" unless end_line && end_column
```

File: lib/spoom/location.rb [hunk 5]
```diff
@@ -15,42 +15,77 @@ class << self
...
+        start_line: T.nilable(Integer),
+        start_column: T.nilable(Integer),
+        end_line: T.nilable(Integer),
+        end_column: T.nilable(Integer),
       ).void
     end
-    def initialize(file, start_line, start_column, end_line, end_column)
+    def initialize(file, start_line: nil, start_column: nil, end_line: nil, end_column: nil)
+      raise LocationError,
+        "Invalid location. End line is required if start line is provided" if start_line && !end_line
```
Comment (Morriar) [inline lib/spoom/location.rb:79]:
```suggestion
        "Invalid location: end line is required if start line is provided" if start_line && !end_line
```

File: lib/spoom/location.rb [hunk 6]
```diff
@@ -15,42 +15,77 @@ class << self
...
+      raise LocationError,
+        "Invalid location. Start line is required if end line is provided" if !start_line && end_line
```
Comment (Morriar) [inline lib/spoom/location.rb:81]:
```suggestion
        "Invalid location: start line is required if end line is provided" if !start_line && end_line
```

File: lib/spoom/location.rb [hunk 7]
```diff
@@ -15,42 +15,77 @@ class << self
...
+      raise LocationError,
+        "Invalid location. End column is required if start column is provided" if start_column && !end_column
```
Comment (Morriar) [inline lib/spoom/location.rb:83]:
```suggestion
        "Invalid location: end column is required if start column is provided" if start_column && !end_column
```

File: lib/spoom/location.rb [hunk 8]
```diff
@@ -15,42 +15,77 @@ class << self
...
+      raise LocationError,
+        "Invalid location. Start column is required if end column is provided" if !start_column && end_column
```
Comment (Morriar) [inline lib/spoom/location.rb:85]:
```suggestion
        "Invalid location: start column is required if end column is provided" if !start_column && end_column
```

File: lib/spoom/location.rb [hunk 9]
```diff
@@ -15,42 +15,77 @@ class << self
...
+      raise LocationError,
+        "Invalid location. Lines are required if columns are provided" if start_column && !start_line
```
Comment (Morriar) [inline lib/spoom/location.rb:87]:
```suggestion
        "Invalid location: lines are required if columns are provided" if start_column && !start_line
```

**Comment by Morriar** [comment]
Indeed, maybe the `Location` concept is not a good representation, maybe what we want is the concept of `File` vs. `FileLocation`. Though I'm wary of doing the refactoring right now as we are still exploring how we want to modelize whole file removals.

For now, the nilability change suits our needs and doesn't commit us to one direction or another 👍

File: lib/spoom/location.rb [hunk 10]
```diff
@@ -15,42 +15,75 @@ class << self
...
+          start_line, end_line = T.must(start_line).split("-", 2)
+          raise LocationError, "Invalid location string `#{location_string}`: missing end line" unless end_line
+
+          return new(file, start_line: start_line.to_i, end_line: end_line.to_i) if end_line
+        end
 
         start_column, rest = rest.split("-", 2)
-        raise LocationError, "Invalid location string: #{location_string}" unless start_column && rest
+        if rest.nil?
+          raise LocationError, "Invalid location string `#{location_string}`: missing end line and column"
```
Comment (Morriar) [inline lib/spoom/location.rb:32]: Does this fit in one line?
```suggestion
        raise LocationError, "Invalid location string `#{location_string}`: missing end line and column" if rest.nil?
```

**Review by Morriar — APPROVED** [review]
One small nit, we can merge once fixed 🎉

File: lib/spoom/location.rb (hunk 10 above)
Comment (alexcrocha) [reply lib/spoom/location.rb:32]: It does, good catch! 🕵️

**Pull request merged**